### ISSUE TITLE
Prefix for nginx proxies?

### ISSUE THREAD

**Issue by KakarottoCake** [MCSManager/MCSManager#885]
I want to add this to my nginx setup.
for example: example.com/funnyminecraftconsole
However, there is no option to enable prefixes in the tool itself, despite being properly added to nginx. Is there a way to do this correctly?

**Comment by bddjr** [comment]
You can change domain, like this:
```
funnyminecraftconsole.example.com
```

***
Related issue:
#826

**Comment by KakarottoCake** [comment]
> You can change domain, like this:
> 
> ```
> funnyminecraftconsole.example.com
> ```
> 
> DNS cname to:
> 
> ```
> example.com
> ```
> 
> Related issue: #826

I apologize, but I speak English. The related issue you linked is completely incomprehensible to me. Is there any chance you can provide the information in there to me in English?

**Comment by unitwk** [comment]
MCSManager does not support this operation.

**Comment by Bluemangoo** [comment]
I mean, only the index html file will be moved, and others won't. in this case, all the requests will go to the right way.

**Comment by Bluemangoo** [comment]
for static files, which server hosts them does not matters, right?

**Comment by huangsijun17** [comment]
MCSManager can be divided into two parts: web and daemon. After I made a certain modification on the web, it seems that it can be reversed to the secondary directory. But there seems to be a problem with online compilation, and I don't have a PR to come in.
https://github.com/huangsijun17/UI/commit/419b0cb28f545d0c4aa165aa6607662b55edd239
The daemon part involves the handshake and authentication of the front and back ends, and I can't modify it. But the daemon does not actually need to use the root directory, and can only reverse the three directories `/socket.io/`, `/upload/` and `/download/`.

**Comment by Bluemangoo** [comment]
Plus, @bddjr wrote a nginx config that merge web and daemon part into one port,  and you might get your idea.

(in chinese) https://github.com/bddjr/nginx-proxy-docs-for-mcsm/blob/master/%E9%85%8D%E7%BD%AEHTTP%E5%8F%8D%E5%90%91%E4%BB%A3%E7%90%86%E4%B8%94%E5%90%88%E5%B9%B6%E7%AB%AF%E5%8F%A3.md

main part of merging:

```conf
location ~ (^/socket.io/)|(^/upload/)|(^/download/) {
            # 填写Daemon进程真正监听的端口号，后面不能加斜杠！
            proxy_pass http://localhost:24444 ;

            # 一些请求头
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header REMOTE-HOST $remote_addr;
            # 用于WebSocket的必要请求头
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
            # 增加响应头
            add_header X-Cache $upstream_cache_status;
            # 禁止客户端缓存，防止客户端未更新内容
            expires -1;
        }
        # 代理Web端
        location / {
            # 填写Web面板端真正监听的端口号
            proxy_pass http://localhost:23333 ;

            # 一些请求头
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header REMOTE-HOST $remote_addr;
            # 用于WebSocket的必要请求头
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
            # 增加响应头
            add_header X-Cache $upstream_cache_status;
            # 禁止客户端缓存，防止客户端未更新内容
            expires -1;
        }
```

**Comment by Bluemangoo** [comment]
socket.io directories **must** be proxied directly, but for web project you can do what you want by modifying source codes, especially in UI project.

**Comment by bddjr** [comment]
或许我们可以换一个思路，统计访问Web端时用到的路径，然后将这些路径指向MCSManager，其它的则指向自己的网站。

这样，只要自己的网站不使用这些路径，就可以进行完美合并。

这个思路的一个明显的缺点是，并不能将 favicon.ico 区分，这将导致图标获取的是自己网站的图标，但这并不影响使用MCSManager。

这个思路或许并不是长久之计，一旦MCSManager为Web端添加了新的路径，将可能导致无法正常访问，需要实时跟进相关更新。

上述的内容是理论，未实测。

**Comment by bddjr** [comment]
刚才说的不用试了，无效的。浏览器访问 `/#/`  路径的时候会自动访问成 `/` 路径。

**Comment by bddjr** [comment]
哦不，我下结论太早了。访问 `/#/` 路径的那部分可以代理成 `/mcsm/` ，浏览器访问时地址栏显示的将是 `/mcsm/#/` 开头。这样就一切正常了

**Comment by bddjr** [comment]
(in chinese)

https://github.com/bddjr/nginx-proxy-docs-for-mcsm/blob/master/通过区分路径实现与网站合并.md

**Comment by Bluemangoo** [comment]
@bddjr #表示书签，在Vue router里面可以在单个页面里面处理多个路径。实际上还是spf。

**Comment by bddjr** [comment]
刚刚简化了一下，下面是最主要的内容
```nginx
        location ~ (^/socket.io/)|(^/upload/)|(^/download/) {
            # 填写Daemon进程真正监听的端口号，后面不能加斜杠！
            proxy_pass http://localhost:24444 ;

            # 一些请求头
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header REMOTE-HOST $remote_addr;
            # 用于WebSocket的必要请求头
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
            # 增加响应头
            add_header X-Cache $upstream_cache_status;
            # 禁止客户端缓存，防止客户端未更新内容
            expires -1;
        }
        # 代理Web端
        location ~ ^(/mcsm$|/mcsm/|/css/|/js/|/static/|/api/|/img/|/fonts/) {
            rewrite /mcsm/ / break; #内部重写为 / 路径
            rewrite /mcsm /mcsm/ redirect; #漏掉斜杠就302重定向到补上斜杠的
            # 填写Web面板端真正监听的端口号，后面不能加斜杠！
            proxy_pass http://localhost:23333 ;

            # 一些请求头
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header REMOTE-HOST $remote_addr;
            # 用于WebSocket的必要请求头
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
            # 增加响应头
            add_header X-Cache $upstream_cache_status;
            # 禁止客户端缓存，防止客户端未更新内容
            expires -1;
        }
```

**Comment by Bluemangoo** [comment]
@bddjr 如果是为了多后端合并，你不应该把css、js等路径直接这样子做。很多后端都需要的。

**Comment by huangsijun17** [comment]
> @bddjr 如果是为了多后端合并，你不应该把css、js等路径直接这样子做。很多后端都需要的。

前端最简单的方法，直接改源码。我PR都推了。

**Comment by bddjr** [comment]
所以我刚开始就建议直接用域名区分，最简单最稳定的办法

**Comment by Bluemangoo** [comment]
May be resolved in #1187. Require tests and feedback [here (with guide)](https://github.com/MCSManager/MCSManager/pull/1187#issuecomment-2042212475).

可能在 #1187 中被解决。[这里](https://github.com/MCSManager/MCSManager/pull/1187#issuecomment-2042212475)要一点测试和反馈（含教程）。
